### ISSUE TITLE
Unfixed security vulnerabilities in vendor code

### ISSUE THREAD

**Issue by danfe** [flacon/alacenc#1]
It's been brought to my attention that macosforge's ALAC code you're using has unfixed [security issues](https://github.com/macosforge/alac/issues/22) which had been fixed upstream on May 23; you might want to merge those.

**Comment by danfe** [comment]
Had there been any progress with this?  Some people are raising concerns that we (FreeBSD) distribute software with known security issues, and call for [strong measures](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=285103).

**Comment by SokoloffA** [comment]
Unfortunately I don't understand this code very well. I just took the Apple code, added cmake support, etc.

I tried to look at the CVEs you cited, but I can't find any useful information. [For example](https://nvd.nist.gov/vuln/detail/CVE-2021-0674), I see that there is some patch with **Patch ID: ALPS06064258**, but where can I see its contents?

**Comment by danfe** [comment]
Yeah, it's kind of confusing.  ALAC's [landing page](https://macosforge.github.io/alac/) points to the original [repo](https://github.com/macosforge/alac) which the issue was filed against, yet it does not contain fixes, except the reference to [one commit](https://github.com/olivierlevon/alac/commit/eb3ae0c).  Perhaps we could merge at least that one and call it a day; if upstream does not consider these bugs serious enough to care for a patch, why should we? :-)

To add a bit of context to the situation, [this comment](https://github.com/mikebrady/shairport-sync/pull/1505#issuecomment-1188800877) also talks about that while the Apple decoder appears to be capable of handling different numbers of channels, bit depths, and rates, it doesn't seem to be maintained upstream any more, and mentions [FFmpeg's ALAC decoder](https://github.com/FFmpeg/FFmpeg/blob/master/libavcodec/alac.c) as an alternative.  I'm not sure whether it's available as ready-to-use standalone library or, if not, how feasible would it be to rip this code from FFmpeg, as pulling the whole beast for just one decoder which is only used during testing seems unfitting.

**Comment by SokoloffA** [comment]
I [merged the patch](https://github.com/flacon/alacenc/commit/7e8d065b7dd405f4be0841fed6887079992029d4). It looks like a profanation, but something is better than nothing.

> which is only used during testing seems unfitting.

There's some confusion here, I don't use it just for tests.  In any case, it seems that we should look for a replacement for this code.

1. Someone can use ffmpeg, but he should remember that ffmpeg is distributed under LGPL/GPL. For some projects it may be unacceptable.
2. I also found a [port of this code on rust](https://github.com/LinusU/rust-alac-encoder).  I have no expertise in rust, so I can't determine its safety. It has 15 unsafe blocks, I don't know if that's a few or a lot.

**Comment by danfe** [comment]
> I [merged the patch](https://github.com/flacon/alacenc/commit/7e8d065b7dd405f4be0841fed6887079992029d4).

Thanks, I'll add it to the FreeBSD port.

> It looks like a profanation

Yeah I know. :-)

> There's some confusion here, I don't use it just for tests.

Sorry, my fault.  I need to review the `ExtProgram::allPrograms()` and update `pkg-message` accordingly, i.e. make it exhaustive.

> I have no expertise in rust, so I can't determine its safety.

Regardless of this code quality, Rust is a big no-no-no due to amount of bloat it entails.  Shall you decide to replace Apple's ALAC implementation, please find an adequate C/C++ one.
